### DISCUSSION TITLE
Solidity Override problem and some confusion

### DISCUSSION THREAD

**Discussion by TreapGoGo** [Cyfrin/foundry-full-course-cu#717] [Q&A] [answered]
I have completed part of the course and been trying to make some stuff on my own.

I tried to do some override practice. However, I got everything messed up.

And here is my code:

```solidity
abstract contract Animal {
    enum Condition {
        ALIVE,
        DEAD
    }
    struct Information {
        uint256 age;
        string name;
        Condition condition;
    }

    Information information;
}

contract Dog is Animal {
    struct Information {
        uint256 age;
        string name;
        uint256 gender;
        Condition condition;
    }

    Information information;
}

abstract contract AnimalMaker {
    function make() external virtual returns (Animal);
}

contract DogMaker is AnimalMaker {
    function make() external override returns (Dog) {
        Dog dog = Dog();
        return dog;
    }
}

```

Now here comes my questions:

1. If the type of return values of sub contract is inherited from the type of return values of parent contract, they are of different types, so the compiler gives me `Overriding function return types differ.` error. How can I solve this?
2. How to override structs and enums? Of course I know that it cannot be overridden. I mean, is there an alternative way to achieve the same goal?
3. When `Dog dog = Dog();` runs, is there a new contract deployed on chain or just staying in memory?

Any help will be appreciated ❤️

**Comment by theirrationalone** [comment, marked as answer]
# Answer 1:

```solidity
contract ContractA {
    struct Animal {
        string name;
        uint256 age;
    }
}

contract ContractB {
    function getAnimalDetail() public view returns (ContractA.Animal memory) {
        return animal; // animal must a type of struct Animal. 
    }
}
```

And remember structs can't be different. 

# Answer2:

```solidity
contract ContractA {
    struct Animal {
        string name;
        uint256 age;
    }
    
    Animal public dog = Animal("limbo", 3);
}

contract ContractB {
    function getAnimalDetail() public view returns (ContractA.Animal memory) {
        return animal; // animal must a type of struct Animal. 
    }
}

contract ContractC {
    function changeVals() public {
        ContractA contA = new ContractA();

        console.log("before:") ;
        console.log(contA.dog().name);
        console.log(contA.dog().age);

        contA.dog = ContractA.Animal("krito", 5);  // true only for `contA` instance of `ContractA`

        console.log("After:") ;
        console.log(contA.dog().name);
        console.log(contA.dog().age);
    }
}
```

We can manipul*** / modify values of an iinstance of a contract, struct, enum, array...

# Answer3: 

`Dog dog = Dog():` ❌

`Dog dog = new Dog();` 🐕 creates a new instance of Dog. 

`Dog dog2 = Dog(address(dog));` 🐕 creates a brand new isolated copy of instance of `dog`

`override` & `virtual` keywords, Using them we can override state variables of anytype through functions who have public Or internal visibility modifiers. 

Sorry if i went wrong somewhere Or somehow... 😅😅

Feel free to correct ✅ me. 😉

Examples above are pseudo codes only for understanding concepts.

**Reply by TreapGoGo** [reply]
Thanks so much for your such a detailed answer! ❤️

**Reply by theirrationalone** [reply]
You're welcome @qpdk777 ! ❤

**Comment by CoggleChucker** [comment]
Hi @qpdk777 
There are few problems in your code.
1>you tried to redefine ```Information``` struct in ```Dog``` contract. This is invalid. If you want, you can add uint256 separate inside Dog contract.
2>In ```DogMaker``` contract you changed return type from ```animal``` to ```dog```. This is invalid. If you are gonna return ```Animal``` type in virtual function, then you have to return the same type in every override function.
3>To create new instance use ```new``` keyword(e.g. ```Dog dog = new Dog();```)

The architecture you are trying to follow here is mostly used in game development.

after tweeking for some time in remix here is what I came up with, for your architecture.

# Step 1:
Base class for all animals
```solidity
//Base Contract
abstract contract Animal
{
    enum Condition {
        ALIVE,
        DEAD
    }

    struct Information {
        uint256 age;
        string name;
        Condition condition;
    }

    Information info;
}
```

# Step 2:
derived classes. Constructors are used to set their values
```solidity
//Inherited Contracts
contract Dog is Animal{
    constructor(uint _age, string memory _name, Condition _condition)
    {
        info = Information(_age,_name,_condition);
    }

    function ChangeAnimalName(string memory _newName) public
    {
        info.name = _newName;
    }
}

contract Cat is Animal{
    constructor(uint _age, string memory _name, Condition _condition)
    {
        info = Information(_age,_name,_condition);
    }
    
    function ChangeAnimalName(string memory _newName) public
    {
        info.name = _newName;
    }
}
```

# Step 3:
Maker contracts:
```solidity
//Maker Contracts
abstract contract AnimalMaker {
    function make() external virtual returns (Animal);
}

contract DogMaker is AnimalMaker {
    function make() external override returns (Animal) {
        Dog dog = new Dog(5, "rocky", Animal.Condition.ALIVE);
        //Add created animals to some kind of list. To keep track of it
        return dog;
    }
}

contract CatMaker is AnimalMaker {
    function make() external override returns (Animal) {
        Cat cat = new Cat(3, "Lili", Animal.Condition.ALIVE);
        //Add created animals to some kind of list. To keep track of it
        return cat;
    }
}
```
In remix IDE, all of the above code successfully compiles. 
I hope it will work for your purposes.

**Reply by TreapGoGo** [reply]
Thanks so much for your such a detailed answer! ❤️ 
Now I come to know that override of solidity is not omnipotent, so sometimes I have to give up reusing some of my code.
